src/main/java/com/google/devtools/build/lib/analysis/skylark/SkylarkActionFactory.java

```diff
@@ -516,13 +518,56 @@ private void registerStarlarkAction(
       builder.useDefaultShellEnvironment();
     }
     if (executionRequirementsUnchecked != Runtime.NONE) {
-      builder.setExecutionInfo(
-          TargetUtils.filter(
-              SkylarkDict.castSkylarkDictOrNoneToDict(
-                  executionRequirementsUnchecked,
-                  String.class,
-                  String.class,
-                  "execution_requirements")));
+      Map<String, String> executionInfo = TargetUtils.filter(
+          SkylarkDict.castSkylarkDictOrNoneToDict(
+              executionRequirementsUnchecked,
+              String.class,
+              String.class,
+              "execution_requirements"));
+      builder.setExecutionInfo(executionInfo);
+
+      String cpuAsString = executionInfo.get(ExecutionRequirements.TAGGED_CPU.tagName());
+      String memoryAsString = executionInfo.get(ExecutionRequirements.TAGGED_MEM_MB.tagName());
+
+      if (cpuAsString != null || memoryAsString != null) {
+        ResourceSet defaultResources = SpawnAction.DEFAULT_RESOURCE_SET;
+        double cpu = defaultResources.getCpuUsage();
+        double memMb = defaultResources.getMemoryMb();
+
+        if (cpuAsString != null) {
+          try {
+            cpu = ExecutionRequirements.TAGGED_CPU.parse(cpuAsString);
+          } catch (ExecutionRequirements.TaggedRequirement.ValidationException e) {
+            throw new EvalException(
+                location,
+                String.format(
+                    "%s has a '%s' tag, but its value '%s' did not pass validation: %s",
+                    ruleContext.getLabel(),
+                    ExecutionRequirements.TAGGED_CPU.tagName(),
+                    e.getTagValue(),
+                    e.getMessage()
+                ));
+          }
+        }
+
+        if (memoryAsString != null) {
```

Comment:
This chunk is a copy/paste of the previous one. Could we deduplicate it behind a function of the form `parseResource(String value, TaggedRequirement<Integer> requirement)` or something?